src/main/java/org/radarbase/appserver/service/scheduler/quartz/MessageJob.java

```diff
@@ -40,67 +47,131 @@
+          break;
+        case DATA:
+          DataMessage dataMessage =
+              dataMessageService.getDataMessageByProjectIdAndSubjectIdAndDataMessageId(
+                  projectId, subjectId, messageId);
+          message = dataMessage;
+          dataMessageSchedulerService.send(dataMessage);
+          break;
+        default:
+          break;
+      }
+    } catch (FirebaseMessagingException exc) {
+      log.error("Error occurred when sending downstream message.", exc);
+      // TODO: update the data message status using event
+      if (message != null) {
+        handleErrorCode(exc.getErrorCode(), message);
+        handleFCMErrorCode(exc.getMessagingErrorCode(), message);
+      }
+    } catch (Exception e) {
+      throw new JobExecutionException(e);
+    }
+  }
+
+  protected void handleErrorCode(ErrorCode errorCode, Message message) {
+    // More info on ErrorCode: https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+    switch (errorCode) {
+      case INVALID_ARGUMENT:
+      case INTERNAL:
+      case ABORTED:
+      case CONFLICT:
+      case CANCELLED:
+      case DATA_LOSS:
+      case NOT_FOUND:
+      case OUT_OF_RANGE:
+      case ALREADY_EXISTS:
+      case DEADLINE_EXCEEDED:
+      case PERMISSION_DENIED:
+      case RESOURCE_EXHAUSTED:
+      case FAILED_PRECONDITION:
+      case UNAUTHENTICATED:
+      case UNKNOWN:
+        break;
+      case UNAVAILABLE:
+        // TODO: Could schedule for retry.
+        log.warn("The FCM service is unavailable.");
+        break;
+    }
+  }
 
-            switch (type) {
-                case NOTIFICATION:
-                    Notification notification =
-                            notificationService.getNotificationByProjectIdAndSubjectIdAndNotificationId(
-                                    projectId, subjectId, messageId);
-                    notificationSchedulerService.send(notification);
-                    break;
-                case DATA:
-                    DataMessage dataMessage =
-                            dataMessageService.getDataMessageByProjectIdAndSubjectIdAndDataMessageId(
-                                    projectId, subjectId, messageId);
-                    dataMessageSchedulerService.send(dataMessage);
-                    break;
-                default:
-                    break;
-            }
-        } catch (Exception e) {
-            throw new JobExecutionException(e);
-        }
+  protected void handleFCMErrorCode(MessagingErrorCode errorCode, Message message) {
+    switch (errorCode) {
+      case INTERNAL:
+      case QUOTA_EXCEEDED:
+      case INVALID_ARGUMENT:
+      case SENDER_ID_MISMATCH:
+      case THIRD_PARTY_AUTH_ERROR:
+        break;
+      case UNAVAILABLE:
+        // TODO: Could schedule for retry.
+        log.warn("The FCM service is unavailable.");
+        break;
+      case UNREGISTERED:
+        FcmUserDto userDto = new FcmUserDto(message.getUser()).setFcmToken("undefined");
```

Comment:
Should the `fcmToken` be maybe a random string or Instant string so it's still unique?

Reply:
hmm good point, yes instant string with a prefix (like `unregistered_<data-time>`) would be unique as well as provide useful information about deregistration.

Reply:
i will wait for the other PR to be merged so i can reuse the function in the service.